src/lib/operators/table_scan.hpp

```diff
@@ -25,7 +28,165 @@ class TableScan : public AbstractOperator {
+      }
+      case ScanType::OpLessThan: {
+        return value_id < lower_bound;
+      }
+      case ScanType::OpLessThanEquals: {
+        if (value_exists) {
+          return value_id <= lower_bound;
+        } else {
+          return value_id < lower_bound;
+        }
+      }
+      case ScanType::OpGreaterThan: {
+        if (value_exists) {
+          return value_id > lower_bound;
+        } else {
+          return value_id >= lower_bound;
+        }
+      }
+      case ScanType::OpGreaterThanEquals: {
+        return value_id >= lower_bound;
+      }
+      default: {
+        throw std::logic_error("Scan type not supported");
+      }
+    }
+  }
+
+  template <typename T>
+  void _collect_matched_rows_for_reference_segment(const std::shared_ptr<ReferenceSegment>& segment,
+                                                   const std::shared_ptr<PosList>& matched_row_ids) const {
+    for (auto const& row_id : *(segment->pos_list())) {
+      const ChunkID chunk_id = row_id.chunk_id;
+      const ChunkOffset real_chunk_offset = row_id.chunk_offset;
+
+      const ColumnID column_id = segment->referenced_column_id();
+      auto table = segment->referenced_table();
+
+      const Chunk& chunk = table->get_chunk(chunk_id);
+      const auto actual_segment = chunk.get_segment(column_id);
+
+      const std::string column_type = table->column_type(_column_id);
+
+      resolve_data_type(column_type, [&](auto type) {
+        using Type = typename decltype(type)::type;
+        const Type search_value = type_cast<Type>(_search_value);
+
+        // handle value segment
+        const auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(actual_segment);
+        if (value_segment) {
+          if (_evaluate_scan_predicate<Type>(search_value, value_segment->values().at(real_chunk_offset))) {
+            matched_row_ids->push_back(RowID{chunk_id, real_chunk_offset});
+          }
+          return;
+        }
+
+        // handle dictionary segment
+        const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(actual_segment);
+        if (dictionary_segment) {
+          if (_evaluate_scan_predicate<Type>(search_value, dictionary_segment->get(real_chunk_offset))) {
+            matched_row_ids->push_back(RowID{chunk_id, real_chunk_offset});
+          }
+          return;
+        }
+        throw std::logic_error("Unknown segment type");
+      });
+    }
+  }
+
+  template <typename T>
+  void _collect_matched_rows_for_value_segment(const ChunkID chunk_id, const std::shared_ptr<ValueSegment<T>>& segment,
+                                               const std::shared_ptr<PosList>& matched_row_ids) const {
+    const T search_value = type_cast<T>(_search_value);
+    ChunkOffset chunk_offset{0};
+    for (const T& value : segment->values()) {
+      if (_evaluate_scan_predicate(search_value, value)) {
+        matched_row_ids->push_back(RowID{chunk_id, chunk_offset});
```

Comment:
`emplace_back` could be faster here:
```suggestion
        matched_row_ids->emplace_back(RowID{chunk_id, chunk_offset});
```